src/compiler/core.ts

```diff
@@ -655,24 +659,39 @@ namespace ts {
         return [...array1, ...array2];
     }
 
-    // TODO: fixme (N^2) - add optional comparer so collection can be sorted before deduplication.
-    export function deduplicate<T>(array: ReadonlyArray<T>, areEqual?: (a: T, b: T) => boolean): T[] {
-        let result: T[];
-        if (array) {
-            result = [];
-            loop: for (const item of array) {
-                for (const res of result) {
-                    if (areEqual ? areEqual(res, item) : res === item) {
-                        continue loop;
-                    }
+    /**
+     * Creates a new array with duplicate entries removed.
+     * @param equalityComparer An optional `EqualityComparer` used to determine if two values are duplicates.
+     * @param comparer An optional `Comparer` used to sort entries before comparison. If supplied,
+     * results are returned in the original order found in `array`.
+     */
+    export function deduplicate<T>(array: ReadonlyArray<T>, equalityComparer?: EqualityComparer<T>, comparer?: Comparer<T>): T[] {
+        if (!array) return undefined;
+        if (!comparer) return addRangeIfUnique([], array, equalityComparer);
+        return deduplicateWorker(array, equalityComparer, comparer);
+    }
+
+    function deduplicateWorker<T>(array: ReadonlyArray<T>, equalityComparer: EqualityComparer<T> = equateValues, comparer: Comparer<T>) {
+        // Perform a stable sort of the array. This ensures the first entry in a list of
+        // duplicates remains the first entry in the result.
+        const indices = sequence(0, array.length);
+        stableSortIndices(array, indices, comparer);
+
+        const deduplicated: number[] = [];
+        loop: for (const sourceIndex of indices) {

```

Comment:
I might be missing something, but it seems like for an array with no duplicates this will do O(n^2) work? We loop over `indices` which has the same length as the input, then we loop over `deduplicated` which will have on average length `n/2`, and with no duplicates the `continue` doesn't matter.